VariantShredding.md

```diff
@@ -25,290 +25,320 @@
+| `["comedy", "drama"]`            | null        | non-null      | [null, null]          | [`comedy`, `drama`]            |
+| `["horror", null]`               | null        | non-null      | [null, `00`]          | [`horror`, null]               |
+| `["comedy", "drama", "romance"]` | null        | non-null      | [null, null, null]    | [`comedy`, `drama`, `romance`] |
+| null                             | `00` (null) | null          |                       |                                |
 
-# Shredding Semantics
+#### Objects
 
-Reconstruction of Variant value from a shredded representation is not expected to produce a bit-for-bit identical binary to the original unshredded value.
-For example, in a reconstructed Variant value, the order of object field values may be different from the original binary.
-This is allowed since the [Variant Binary Encoding](VariantEncoding.md#object-field-id-order-and-uniqueness) does not require an ordering of the field values, but the field IDs will still be ordered lexicographically according to the corresponding field names.
+Fields of an object can be shredded using a Parquet group for `typed_value` that contains shredded fields.
 
-The physical representation of scalar values may also be different in the reconstructed Variant binary.
-In particular, the [Variant Binary Encoding](VariantEncoding.md) considers all integer and decimal representations to represent a single logical type.
-This flexibility enables shredding to be applicable in more scenarios, while maintaining all information and values losslessly.
-As a result, it is valid to shred a decimal into a decimal column with a different scale, or to shred an integer as a decimal, as long as no numeric precision is lost.
-For example, it would be valid to write the value 123 to a Decimal(9, 2) column, but the value 1.234 would need to be written to the `variant_value` column.
-When reconstructing, it would be valid for a reader to reconstruct 123 as an integer, or as a Decimal(9, 2).
-Engines should not depend on the physical type of a Variant value, only the logical type.
+If the value is an object, `typed_value` must be non-null.
+If the value is not an object, `typed_value` must be null.
+Readers can assume that a value is not an object if `typed_value` is null and that `typed_value` field values are correct; that is, readers do not need to read the `value` column if `typed_value` fields satisfy the required fields.
 
-On the other hand, shredding as a different logical type is not allowed.
-For example, the integer value 123 could not be shredded to a string `typed_value` column as the string "123", since that would lose type information.
-It would need to be written to the `variant_value` column.
+Each shredded field in the `typed_value` group is represented as a required group that contains optional `value` and `typed_value` fields.
```

Comment:
Are there any cases where a field's name might not be a valid Parquet group name, making it impossible to shred? E.g. are there any restrictions on whitespace or non-printable characters in Parquet? Or any that should be discouraged for shredding, because readers are known to have trouble with them?

Reply:
There are no specific restrictions in the spec that I can recall, there are certainly engines that don't support some characters but shredded columns should hopefully never reach engine schemas since only the top level variant would ideally be displayed.  For non-displayable characters general display and SQL projection functions would likely become cumbersome.

Some reference implementations might have issues with periods (`.`) but from an API perspective it isn't clear to me if there should be a separate API anyways to interrogate if certain variants are shredded.  This might be a design question to raise on the mailing list before beginning implementation.